Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,338 @@
+Authors: Adrian Wüthrich
+-/
+import Mathlib.Combinatorics.SimpleGraph.AdjMatrix
+import Mathlib.Combinatorics.SimpleGraph.IncMatrix
+import Mathlib.LinearAlgebra.Finrank
+import Mathlib.LinearAlgebra.Matrix.PosDef
+import Mathlib.LinearAlgebra.Matrix.IsDiag
+
+/-!
+# Laplacian Matrix
+
+This module defines the Laplacian matrix of a graph, and proves some of its elementary properties.
+
+## Main definitions & Results
+
+* `SimpleGraph.degMatrix`: The degree matrix of a simple graph
+* `SimpleGraph.lapMatrix `: The Laplacian matrix of a simple graph, defined as the difference
+  between the degree matrix and the adjacency matrix.
+* `isPosSemidef_lapMatrix`: The Laplacian matrix is positive semidefinite.
+* `rank_ker_lapMatrix_eq_card_ConnectedComponent`: The number of connected components in `G` is
+  the dimension of the nullspace of its Laplacian matrix.
+
+-/
+
+
+open BigOperators Finset Matrix SimpleGraph
+
+variable {V α : Type*} (α) [Fintype V] [DecidableEq V] (G : SimpleGraph V) [DecidableRel G.Adj]
+
+/-- The diagonal matrix consisting of the degrees of the vertices in the graph. -/
+def SimpleGraph.degMatrix [AddMonoidWithOne α] : Matrix V V α := Matrix.diagonal (G.degree ·)
+
+/-- `lapMatrix G R` is the matrix `L = D - A` where `D`is the degree
+  and `A` the adjacency matrix of `G`. -/
+def SimpleGraph.lapMatrix [AddGroupWithOne α] : Matrix V V α := G.degMatrix α - G.adjMatrix α
+
+variable {α}
+
+theorem transpose_lapMatrix [AddGroupWithOne α] : (G.lapMatrix α)ᵀ = G.lapMatrix α := by
+  unfold lapMatrix degMatrix
+  rw [transpose_sub, diagonal_transpose, transpose_adjMatrix]
+
+theorem isSymm_lapMatrix [AddGroupWithOne α] : (G.lapMatrix α).IsSymm := transpose_lapMatrix G
+
+theorem lapMatrix_mulVec_const_eq_zero [Ring α] :
+    mulVec (G.lapMatrix α) (Function.const V 1) = 0 := by
+  unfold lapMatrix
+  rw [sub_mulVec]
+  ext
+  simp only [Pi.const_one, Pi.sub_apply, adjMatrix_mulVec_apply, Pi.one_apply, sum_const,
+    card_neighborFinset_eq_degree, nsmul_eq_mul, mul_one, Pi.zero_apply]
+  unfold mulVec dotProduct
+  simp only [Pi.one_apply, mul_one]
+  unfold degMatrix diagonal
+  simp only [of_apply, sum_ite_eq, mem_univ, ite_true, sub_self]
+
+theorem dotProduct_mulVec_adjMatrix [Ring α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.adjMatrix α) x = ∑ i : V, ∑ j : V, if G.Adj i j then x i * x j else 0 := by
+  unfold dotProduct mulVec dotProduct
+  simp [mul_sum]
+
+theorem dotProduct_mulVec_degMatrix [CommRing α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.degMatrix α) x = ∑ i : V, G.degree i * x i * x i := by
+  unfold dotProduct degMatrix
+  simp only [mulVec_diagonal, ← mul_assoc, mul_comm (x _)]
+
+variable (α)
+
+theorem sum_adj_eq_degree [AddCommMonoidWithOne α] (i : V) :
+    (G.degree i : α) = ∑ j : V, if G.Adj i j then 1 else 0 := by
+  rw [sum_boole]
+  unfold degree neighborFinset neighborSet
+  rw [Set.toFinset_setOf fun x ↦ Adj G i x]
+
+theorem ite_sub_distr {α : Type u_1} [NonAssocRing α] (P : Prop) [Decidable P] (a b : α) :
```

Comment:
This theorem and the one below should be in a much earlier file. Maybe in Mathlib.Algebra.Ring.Defs next to [ite_add](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Ring/Defs.html#ite_add)

Reply:
No actually, `P` does not need multiplication! This should work with just `AddGroup` requirement. Maybe move these to Mathlib.Algebra.Group.Basic ?

Reply:
I renamed both theorems and made them more general, I hope this is better:
```lean
theorem ite_sub_ite {α : Type*} [Sub α] (P : Prop) [Decidable P] (a b c d : α) :
    ((if P then a else b) - if P then c else d) = if P then a - c else b - d := by
  split
  repeat rfl

theorem ite_add_ite {α : Type*} [Add α] (P : Prop) [Decidable P] (a b c d : α) :
    ((if P then a else b) + if P then c else d) = if P then a + c else b + d := by
  split
  repeat rfl
```
I'm not quite sure where to put it in `AddGroup`. Also `ite_add` does not need multiplication either, why is it in Mathlib.Algebra.Ring.Defs? Finally if I put these two theorems in another file, should I create a new pull request?

Reply:
Good point. Mathlib.Algebra.Ring.Defs sounds good then. 

No, you don't need a new PR. It's ok to have multiple files being edited in a single PR.

Reply:
This PR is really cool btw. I'm looking forward to it getting merged and having more Spectral Graph Theory in mathlib.

Reply:
Got it done now